signature/internal/sequoia/rust/src/signature.rs

```diff
@@ -0,0 +1,433 @@
+
+        let mut keys = self.keystore.find_key(signing_key_handles[0].clone())?;
+
+        if keys.is_empty() {
+            return Err(anyhow::anyhow!("No matching key in keystore"));
+        }
+        if let Some(password) = password {
+            keys[0].unlock(password.into())?;
+        }
+
+        let mut sink = vec![];
+        {
+            let message = Message::new(&mut sink);
+            let message = Signer::new(message, &mut keys[0])?.build()?;
+            let mut message = LiteralWriter::new(message).build()?;
+            message.write_all(data)?;
+            message.finalize()?;
+        }
+        Ok(sink)
+    }
+
+    fn verify(&mut self, signature: &[u8]) -> Result<SequoiaVerificationResult, anyhow::Error> {
+        if signature.is_empty() {
+            return Err(anyhow::anyhow!("empty signature"));
+        }
+
+        let h = Helper {
+            certstore: self.certstore.clone(),
+            signer: Default::default(),
+        };
+        let mut policy = ConfiguredStandardPolicy::new();
+        policy.parse_default_config()?;
+        let policy = policy.build();
+
+        let mut v = VerifierBuilder::from_bytes(signature)?.with_policy(&policy, None, h)?;
+        let mut content = Vec::new();
+        v.read_to_end(&mut content)?;
+
+        assert!(v.message_processed());
+
+        match &v.helper_ref().signer {
+            Some(signer) => Ok(SequoiaVerificationResult {
+                content,
+                signer: CString::new(signer.fingerprint().to_hex().as_bytes()).unwrap(),
+            }),
+            None => Err(anyhow::anyhow!("No valid signature")),
+        }
+    }
+}
+
+struct Helper<'a> {
+    certstore: Arc<sequoia_cert_store::CertStore<'a>>,
+    signer: Option<openpgp::Cert>,
+}
+
+impl<'a> VerificationHelper for Helper<'a> {
+    fn get_certs(&mut self, ids: &[openpgp::KeyHandle]) -> openpgp::Result<Vec<openpgp::Cert>> {
+        let mut certs = Vec::new();
+        for id in ids {
+            let matches = self.certstore.lookup_by_cert_or_subkey(id);
+            for lc in matches? {
+                certs.push(lc.to_cert()?.clone());
+            }
+        }
+        Ok(certs)
+    }
+
+    fn check(&mut self, structure: MessageStructure) -> openpgp::Result<()> {
+        for layer in structure {
+            match layer {
+                MessageLayer::Compression { algo } => log::info!("Compressed using {}", algo),
+                MessageLayer::Encryption {
+                    sym_algo,
+                    aead_algo,
+                } => {
+                    if let Some(aead_algo) = aead_algo {
+                        log::info!("Encrypted and protected using {}/{}", sym_algo, aead_algo);
+                    } else {
+                        log::info!("Encrypted using {}", sym_algo);
+                    }
+                }
+                MessageLayer::SignatureGroup { ref results } => {
+                    let result = results.iter().find(|r| r.is_ok());
```

Comment:
It would really help diagnosing failures if the errors from `results` were returned to the caller, and not replaced by a generic `No valid signature`.

Reply:
Attempted in https://github.com/containers/image/pull/2876 .